Campaign refactor

nlukic97 opened the pull request:
We have made many changes to the existing contracts, as well as adding a new contract.
Please ignore the contracts found in the folders `./contracts/airdropsSMCampaign-old`. They are our legacy campaign contracts which we kept for reference purposes. Also, please ignore any contracts which have a coverage of 0, as these are also legacy contracts we have been told to keep.

bxmmm1 commented:
Create solcover config https://github.com/sc-forks/solidity-coverage#config-options because it will make coverage look nice, and also if you decide to introduce CI it will not look like coverage is lower than it really is because of some reference contracts.

bxmmm1 commented on .gitignore:
```diff
@@ -15,6 +15,7 @@
 **/node_modules
 .idea/*
+.vscode
 
```
add .DS_Store to gitignore

nlukic97 replied:
I have had .DS_Store show up sometimes in the project (and I would manually delete it), but sometimes it doesn't show up. What causes this to be generated?

bxmmm1 replied:
Macbook automatically generates those files, idk the reason never really researched it

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -68,69 +53,46 @@ contract Existing1155NftDrop is AirdropInfo, AirdropMerkleProof, IERC1155Receive
-    function batchClaim(uint256[] memory tokenIds, bytes32[] calldata _merkleProof) external {
-        if (rewardToken.balanceOf(address(this), rewardTokenId) < tokensPerClaim * tokenIds.length)
-            revert InsufficientLiquidity();
+    function batchClaim(uint256[] memory tokenIds) external {
```
natspec missing

bxmmm1 replied:
As a rule of thumb you want to use `calldata` when you're not planing to mutate array.
In this case you are only reading data and not mutating, so using `calldata` would be a gas optimization.

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -68,69 +53,46 @@ contract Existing1155NftDrop is AirdropInfo, AirdropMerkleProof, IERC1155Receive
         airdropFundBlockTimestamp = block.timestamp;
         airdropFundingHolder = msg.sender;
-        emit AirdropFunded();
+        emit AirdropFunded(address(this));
```
I still don't get this, pls comment when you find out from frontend team

nlukic97 replied:
The developer who requested this is on holiday, will check with them first thing on monday

nlukic97 replied:
No update on this from the backend dev who requested it. I suggest we merge it and update when more information is available.

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -139,40 +101,24 @@ contract Existing1155NftDrop is AirdropInfo, AirdropMerkleProof, IERC1155Receive
-    //@notice Checks if the user is eligible for this airdrop
-    function isEligibleForReward(uint256 tokenId) external view returns (bool) {
+    /// @notice Checks if the user is eligible for this airdrop
+    function isEligibleForReward(uint256 tokenId) public view returns (bool) {
```
natspec @return missing

nlukic97 replied:
Just to check, would this be appropriate?
<img width="848" alt="image" src="https://user-images.githubusercontent.com/56925545/180454769-aab124ed-3a45-4fca-a3fe-e317cec9e3a8.png">

Since

nlukic97 replied:
Also, we have not added the natspec for return value anywhere in this project so we'd have to fix that too

bxmmm1 replied:
You can do it like this, or `@return true if user is eligible for reward` or something like that. 
It is a good practice to see well-documented code, unless you are in a hurry to deploy, I'd make it right

bxmmm1 replied:
I'd change naming a bit for this, or the way it is written.

The function is called `isEligibleForReward` and returns a bool
Without looking at the code, and just looking at this line `function isEligibleForReward(uint256 tokenId) public view returns (bool)`
I'd expect this function to return true/false
Because this function never returns false, it wastes gas by returning `true`.

In our case we never return `false`, instead we revert for some reason.
If you want to keep the code this way I recommend renaming to something like `checkRewardEligibility` or `validateRewardEligibility`, or something like that.
Imagine that you're looking at the code that somebody wrote and you see `isOver18`, You wouldn't expect a revert if somebody is 15, `false` makes more sense.

This is not a major issue, just the way I think when i read function signatures.

nlukic97 replied:
Fixed the natspec's and addid a validation function, whilst making sure the isEligibleForReward function only returns true or false.

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -139,40 +101,24 @@ contract Existing1155NftDrop is AirdropInfo, AirdropMerkleProof, IERC1155Receive
+    /// @notice Checks if the user is eligible for this airdrop
+    function isEligibleForReward(uint256 tokenId) public view returns (bool) {
         if (hasClaimed[tokenId]) revert AlreadyRedeemed();
+        if (rewardedNft.ownerOf(tokenId) != msg.sender) revert Unauthorized();
```
I don't understand this.. 
User is supposed to find out if he is eligible for reward, 
and here we are checking if he owns `rewardedNft` and revert if he doesn't

nlukic97 replied:
In order to be eligible for reward, the user must own the rewardedNfr

bxmmm1 replied:
My bad, brain afk on friday :))

nlukic97 replied:
No worries, 17:00 is nearly here! 🎉

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -68,69 +53,46 @@ contract Existing1155NftDrop is AirdropInfo, AirdropMerkleProof, IERC1155Receive
 
             if (hasClaimed[tokenId]) revert AlreadyRedeemed();
-            if (rewardedNft.ownerOf(tokenId) != msg.sender) revert NotOwner();
+            if (rewardedNft.ownerOf(tokenId) != msg.sender) revert Unauthorized();
```
see line 108 comment please

nlukic97 replied:
I'm not sure If understand what you are referring to

bxmmm1 replied:
Forget that I wrote it

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -3,59 +3,44 @@ pragma solidity ^0.8.14;
-import "@openzeppelin/contracts/access/Ownable.sol";
-import "../interfaces/AirdropInfo.sol";
 import "../interfaces/AirdropMerkleProof.sol";
 import "../interfaces/IAirBroFactory.sol";
```
unused import

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -68,69 +53,49 @@ contract Existing1155NftDrop is AirdropInfo, AirdropMerkleProof, IERC1155Receive
+            rewardToken.safeTransferFrom(address(this), msg.sender, rewardTokenId, tokensPerClaim, "");
+            emit Claimed(tokenId, msg.sender);
+        } else {
+            revert NotEligible();
```
Please see comment on line 109 regarding `isEligibleForReward` With current code we will never get to this revert because `isEligibleForReward` never returns false

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -64,59 +48,52 @@ contract ExistingTokenDrop is AirdropInfo, AirdropMerkleProof, Ownable {
         airdropFundingHolder = msg.sender;
-        emit AirdropFunded();
+
+        rewardToken.transferFrom(msg.sender, address(this), totalAirdropAmount);
```
Use SafeERC20 for ERC20 tokens that you don't control. Some of them don't revert, instead they return `false` so in that case, the token transfer would fail, but this would not revert making the contract useless, because it would be considered funded, and it would contain no token

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -64,59 +48,52 @@ contract ExistingTokenDrop is AirdropInfo, AirdropMerkleProof, Ownable {
+    function claim(uint256 tokenId) external {
+        if (isEligibleForReward(tokenId)) {
+            hasClaimed[tokenId] = true;
+            rewardToken.transfer(msg.sender, tokensPerClaim);
```
safeTransfer

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -125,34 +102,19 @@ contract ExistingTokenDrop is AirdropInfo, AirdropMerkleProof, Ownable {
         rewardToken.transfer(msg.sender, tokensPerClaim * tokenIds.length);
```
safeTransfer

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -64,59 +48,52 @@ contract ExistingTokenDrop is AirdropInfo, AirdropMerkleProof, Ownable {
-    function batchClaim(uint256[] memory tokenIds, bytes32[] calldata _merkleProof) external {
+    /// @notice Claim multiple ERC20 airdrops at once
+    /// @param tokenIds are the rewarded NFT collections token ID's
+    function batchClaim(uint256[] memory tokenIds) external {
```
calldata, see comment contracts/airdrops/Existing1155NftDrop.sol regarding explanation

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -125,34 +102,19 @@ contract ExistingTokenDrop is AirdropInfo, AirdropMerkleProof, Ownable {
+    /// @notice Checks if the user is eligible for this airdrop
+    /// @param tokenId is the rewarded NFT token ID
+    /// @return bool if user is eligible for reward
+    function isEligibleForReward(uint256 tokenId) public view returns (bool) {
```
see comment on contracts/airdrops/Existing1155NftDrop.sol

bxmmm1 commented on contracts/airdrops/ItemNFTDrop.sol:
```diff
@@ -107,16 +99,4 @@ contract ItemNFTDrop is ERC1155, AirdropInfo, AirdropMerkleProof, Ownable {
     function getAirdropAmount() external view returns (uint256) {
```
/// for natspec

bxmmm1 commented on contracts/airdrops/ItemNFTDrop.sol:
```diff
@@ -90,11 +87,6 @@ contract ItemNFTDrop is ERC1155, AirdropInfo, AirdropMerkleProof, Ownable {
-    }
-
     //@notice Checks if the user is eligible for this airdrop
     function isEligibleForReward(uint256 tokenId) external view returns (bool) {
```
/// and @return

nlukic97 replied:
This is also a deprecated contract, keeping it just in case

bxmmm1 commented on contracts/airdrops/NFTDrop.sol:
```diff
@@ -117,16 +110,4 @@ contract NFTDrop is ERC721, AirdropInfo, AirdropMerkleProof, Ownable {
     function getAirdropAmount() external view returns (uint256) {
```
///

nlukic97 replied:
This is actually a deprecated contract, we are keeping it for a "just in case" scenario so we're not really working on it

bxmmm1 commented on contracts/airdrops/NFTDrop.sol:
```diff
@@ -100,11 +98,6 @@ contract NFTDrop is ERC721, AirdropInfo, AirdropMerkleProof, Ownable {
-    }
-
     //@notice Checks if the user is eligible for this airdrop
     function isEligibleForReward(uint256 tokenId) external view returns (bool) {
```
///, same applies as comments in contracts/airdrops/Existing1155NftDrop.sol

bxmmm1 commented on contracts/airdrops/TokenDrop.sol:
```diff
@@ -4,85 +4,63 @@ pragma solidity ^0.8.14;
 
 /// @title Airdrops new ERC20 tokens for airdrop recipients
-contract TokenDrop is ERC20, AirdropInfo, AirdropMerkleProof, Ownable {
+contract TokenDrop is ERC20, AirdropMerkleProof {
```
For new ERC20, I'd use https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20Permit
as they have gas-less approvals with EIP712 signatures

nlukic97 replied:
Just gonna leave some useful links for later reference
- [ERC20Permin on OpenZeppelin](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#ERC20Permit-nonces-address-)
- [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612)

bxmmm1 commented on contracts/airdrops/TokenDrop.sol:
```diff
@@ -4,85 +4,63 @@ pragma solidity ^0.8.14;
+    uint256 public immutable airdropFinishTime;
+
+    mapping(uint256 => bool) public hasClaimed;
+    string public airdropType = "ERC20";
```
constant

bxmmm1 commented on contracts/airdrops/NFTDrop.sol:
```diff
@@ -3,26 +3,24 @@ pragma solidity ^0.8.14;
     address public immutable airBroFactoryAddress;
 
     string public baseURI;
+    string airdropType = "ERC721";
```
constant

bxmmm1 commented on contracts/airdrops/ItemNFTDrop.sol:
```diff
@@ -3,19 +3,17 @@ pragma solidity ^0.8.14;
     address public immutable airBroFactoryAddress;
 
     string public baseURI;
+    string airdropType = "ERC1155";
```
constant

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -3,59 +3,43 @@ pragma solidity ^0.8.14;
 
-    bool public airdropFunded = false;
+    mapping(uint256 => bool) public hasClaimed;
+    string public airdropType = "ERC20";
```
constant

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -4,61 +4,46 @@ pragma solidity ^0.8.14;
 
-    bool public airdropFunded = false;
+    mapping(uint256 => bool) public hasClaimed;
+    string public airdropType = "ERC1155";
```
constant

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -4,61 +4,46 @@ pragma solidity ^0.8.14;
 
 /// @title Airdrops existing ERC1155 tokens for airdrop recipients
-contract Existing1155NftDrop is AirdropInfo, AirdropMerkleProof, IERC1155Receiver, Ownable {
+contract Existing1155NftDrop is AirdropMerkleProof, IERC1155Receiver {
```
Since in all airdrop contract share a lot of data in constructor you can make an abstract contract with a constructor like this, and then have contrats extend it
```
        rewardedNft = IERC721(_rewardedNft);
        tokensPerClaim = _tokensPerClaim;
        totalAirdropAmount = _totalAirdropAmount;
        airdropDuration = _airdropDuration * 1 days;
        airdropStartTime = block.timestamp;
        airdropFinishTime = block.timestamp + airdropDuration;
```
(pls double what data is shared)

bxmmm1 replied:
No need for `AirdropMerkleProof`? you aren't checking proof anywhere

nlukic97 replied:
Created two abstract parent contracts [(see commit here)](https://github.com/MVPWorkshop/Airbro-contracts/commit/9d696e0432bb4ea81131bca735bed7332aad778f) for shared state constructor initialisation. 

Looking for more room to create more shared contracts like this.

Also removed AirdropMerkleProof from all contracts that do not implement it.

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,126 @@
+    address public immutable airbroCampaignFactoryAddress;
+    uint256 public immutable tokenSupply;
+
+    string public airdropType = "ERC20";
```
const / remove
Idk what is the purpose of this hardcoded string

nlukic97 replied:
It's basically a way for users to know what the reward will be. Inherited from previous developer so we kept it

bxmmm1 replied:
we can change it to constant then, as it won't change

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,126 @@
+
+        merkleRootSet = true;
+        merkleRoot = _merkleRoot;
+        airdropExpirationTimestamp = block.timestamp + 60 days;
```
Maybe constructor is a better place for this + it be adjusted and immutable to save some gas

nlukic97 replied:
What are you referring to specifically?

bxmmm1 replied:
I kinda see it that Admin will deploy this contract, and then setMerkleRoot the same day / or shortly after. It would be cheaper to initialize this `airdropExpirationTimestamp` in constructor like this `airdropExpirationTimestamp = block.timestamp + 60 days;`
OR
have it as construcotr param `_airdropExpirationTimestamp`
and set it when deploying to whatever you want, 5 days, 30 days, 60 days 
That way it looks more flexibile to me, maybe somebody doesn't want 60 days expiration time?

nlukic97 replied:
So basically the merkleRoot is set when the backend decides the campaign is finished, it might happen 1 month after creating the campaign. 

After setting it, all users who have a valid merkleProof have 60 days to claim their reward. 

If they do not claim this reward within 60 days, the funder of the contract (might be the admin but also might not be), can withdraw any remaining funds from the contract, this is to avoid permanently locking funds if someone doesn't withdraw the rewad.

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,126 @@
+        numberOfClaimers = _numberOfClaimers;
+        tokensPerClaim = tokenSupply / numberOfClaimers;
+
+        emit MerkleRootChanged(_merkleRoot, _numberOfClaimers);
```
`MerkleRootChanged` -> `MerkleRootSet` makes more sense to me, as function is called `setMerkleRoot`

nlukic97 replied:
Agreed. Also the MerkleRoot can only be set once, so it makes sense to change the event to match this assumption. Good catch !

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,126 @@
+    function fundAirdrop() external {
+        if (airdropFunded) revert AlreadyFunded();
+
+        if (rewardToken.balanceOf(msg.sender) < tokenSupply) revert InsufficientAmount();
```
`safeTransferFrom` on line 80 will revert if the msg.sender doesn't have the tokens / did not approve them.

nlukic97 replied:
We perform checks for allowance and balance (line 79 and 80) prior to the safeTransfer on line 85, so this should be okay

bxmmm1 replied:
See comment for line 101

nlukic97 replied:
safeTransfer already does the checks, variables set will revert if **safeTransfer** reverts

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,126 @@
+        airdropFunder = msg.sender;
+
+        rewardToken.safeTransferFrom(msg.sender, address(this), tokenSupply);
+        emit AirdropFunded(address(this));
```
address(this) doesn't make sense to me

nlukic97 replied:
Requested feature from a frontend dev, still waiting to confirm the reason behind it

nlukic97 replied:
No update on this from the backend dev who requested it. I suggest we merge it and update when more information is available.

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,126 @@
+    /// @param _merkleProof is the merkle proof that this user is eligible for claiming the ERC20 airdrop
+    function claim(bytes32[] calldata _merkleProof) external {
+        if (isEligibleForReward(_merkleProof)) {
+            if (rewardToken.balanceOf(address(this)) < tokensPerClaim) revert InsufficientLiquidity();
```
safeTransfer will fail on 104 if you remove this line

nlukic97 replied:
I'm not sure I understand the feedback / our mistake, could you please elaborate on what the mistake is?

bxmmm1 replied:
If you try `rewardToken.safeTransfer(msg.sender, tokenAmount);`
And `msg.sender`  doesn't have enough `tokenAmount` this call `rewardToken.safeTransfer(msg.sender, tokenAmount);` will revert.
So why double check something that is already checked in `safeTransfer`

Backend / Subgraph can monitor smart contract, and return data to frontend if there is not enough liquidity.
Here we're doing:
Balance checking before transfer
Transfer
    transfer internally is checking balance and reverting if fails

nlukic97 replied:
Ok, so:
safeTransfer already does the checks, variables set will revert if **safeTransfer** reverts. With revert, no state is changed

bxmmm1 commented on contracts/campaignAirdrops/NewERC1155DropCampaign.sol:
```diff
@@ -0,0 +1,73 @@
+/// @title Airdrops existing ERC1155 tokens for airdrop recipients
+contract NewERC1155DropCampaign is ERC1155, AirdropMerkleProof {
+    address public immutable airbroCampaignFactoryAddress;
+    uint8 private immutable _tokenId = 0; // yes, for immutable/constant I had to set the variable to = 0;
```
constant and not immutable, you want immutable when initializing in constructor

bxmmm1 replied:
For constants and immutables it is better to use uint256 than uint8
Feel free to copy to remix and test deployment cost
```
pragma solidity 0.8.15;

// 80231
contract Asm {    
    uint256 public constant asd = 1;
}

// 81095
contract Asm2 {    
    uint8 public constant asd = 1;
}
```

nlukic97 replied:
@GotchaGuy 
Balance after Asm deployment: 99.99999999999990**8125**
Balance after Asm2 deployment: 99.99999999999990**7477**

Gas saved with Asm2: **648**

bxmmm1 commented on contracts/campaignAirdrops/NewERC1155DropCampaign.sol:
```diff
@@ -0,0 +1,73 @@
+contract NewERC1155DropCampaign is ERC1155, AirdropMerkleProof {
+    address public immutable airbroCampaignFactoryAddress;
+    uint8 private immutable _tokenId = 0; // yes, for immutable/constant I had to set the variable to = 0;
+    uint8 private immutable _tokenAmount = 1;
```
constant and not immutable, you want immutable when initializing in constructor

bxmmm1 commented on contracts/campaignAirdrops/NewERC1155DropCampaign.sol:
```diff
@@ -0,0 +1,73 @@
+
+        merkleRootSet = true;
+        merkleRoot = _merkleRoot;
+        emit MerkleRootChanged(_merkleRoot);
```
MerkleRootSet

bxmmm1 commented on contracts/campaignAirdrops/NewERC1155DropCampaign.sol:
```diff
@@ -0,0 +1,73 @@
+
+    /// @notice Checks if the user is eligible for this airdrop
+    /// @param _merkleProof is the merkleRoot proof that this user is eligible for claiming reward
+    function isEligibleForReward(bytes32[] calldata _merkleProof) public view returns (bool) {
```
same comment applies that I wrote on existing1155nftdrop regarding true/false and reverts

nlukic97 replied:
So: **isEligiblibleForReward** does revert instead of returning bool which might be expected behavior

bxmmm1 commented on contracts/interfaces/AirdropCampaignInfo.sol:
```diff
@@ -0,0 +1,12 @@
+
+    /// @notice Returns the amount(number) of airdrop tokens to claim
+    /// @param _merkleProof is the rewarded NFT collections token ID
+    function getAirdropAmount(bytes32[] calldata _merkleProof) external returns (uint256);
```
/// @return

bxmmm1 commented on contracts/interfaces/AirdropCampaignInfo.sol:
```diff
@@ -0,0 +1,12 @@
+interface AirdropCampaignInfo {
+    /// @notice Checks if the user is eligible for this airdrop
+    /// @param _merkleProof is the merkleRoot proof that this user is eligible for claiming reward
+    function isEligibleForReward(bytes32[] calldata _merkleProof) external returns (bool);
```
/// @return

bxmmm1 commented on contracts/AirbroCampaignFactory.sol:
```diff
@@ -0,0 +1,63 @@
+contract AirbroCampaignFactory {
+    // index of deployed airdrop contracts
+    address[] public airdrops;
+    address public admin = 0xF4b5bFB92dD4E6D529476bCab28A65bb6B32EFb3;
```
maybe set in constructor, it gives more flexibility for writing tests and everything

nlukic97 replied:
Good call, will update

bxmmm1 commented on contracts/AirbroCampaignFactory.sol:
```diff
@@ -0,0 +1,63 @@
+
+    uint256 public totalAirdropsCount;
+
+    event NewAirdrop(address indexed airdropContract, address indexed airdropCreator);
```
Probably it would be good to know if it is 1155 or 20 airdrop in this event

bxmmm1 commented on contracts/airdrops/TokenDrop.sol:
```diff
@@ -4,85 +4,63 @@ pragma solidity ^0.8.14;
-import "@openzeppelin/contracts/access/Ownable.sol";
-import "../interfaces/AirdropInfo.sol";
 import "../interfaces/AirdropMerkleProof.sol";
 import "../interfaces/IAirBroFactory.sol";
```
unused import?

bxmmm1 commented on contracts/AirbroFactorySMCampaign.sol:
```diff
@@ -0,0 +1,140 @@
+        emit NewAirdrop(rewardedNftCollection, address(tokenDropContract), msg.sender);
+    }
+
+    // newERC1155Drop, ExistingERC20Drop
```
?

bxmmm1 commented on contracts/AirbroFactorySMCampaign.sol:
```diff
@@ -0,0 +1,140 @@
+    // index of deployed airdrop contracts
+    address[] public airdrops;
+    address[] public nft1155Contracts;
+    address public admin = 0xF4b5bFB92dD4E6D529476bCab28A65bb6B32EFb3;
```
constructor?

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+    /// @param _airdropCampaignAddress - address of the airdropCampaign contract whose current participants are in the daily merkleRootHash
+    /// @param _merkleRootHash - merkle root hash of daily participants of an airdropCampaign
+    function addDailyMerkleRootHash(address _airdropCampaignAddress, bytes32 _merkleRootHash) external onlyAdmin {
+        if (airdrops[_airdropCampaignAddress].airdropFinished == true) revert AirdropHasFinished();
```
no need to check == true you can just `if (airdrops[_airdropCampaignAddress].airdropFinished)

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+    /// @notice Adds array of daily merkle root hashes for multiple airdropCampaign contracts
+    /// @param _airdropCampaignAddressArray - array of addresses of multiple airdropCampaign contracts whose current participants are in the daily merkleRootHash
+    /// @param _merkleRootHashArray - array of merkle root hashes of daily participants of multiple airdropCampaigns
+    function batchAddDailyMerkleRootHash(address[] memory _airdropCampaignAddressArray, bytes32[] memory _merkleRootHashArray)
```
calldata for _airdropCampaignAddressArray

nlukic97 replied:
Applied it for this and for _merkleRootHashArray 👍

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+        onlyAdmin
+    {
+        uint256 airdropHashArrayLength = _merkleRootHashArray.length;
+        if (airdropHashArrayLength > batchHashArrayLimit) revert ArrayTooLong();
```
not needed, line 71 covers this case

nlukic97 replied:
Line 71 just checks if they are the same length, but line 70 checks if the length on one does not exceed 600. 

I'll speak to the team about removing this limit we have set

nlukic97 replied:
Edit: removed this check

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+            bytes32 currentMerkleRootHash = _merkleRootHashArray[i];
+
+            if (airdrops[currentCampaignAddress].chain == Chains.Zero) revert ChainDataNotSet();
+            if (airdrops[currentCampaignAddress].airdropFinished == true) revert AirdropHasFinished();
```
no need to check == true you can just if(..)

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+    function addAirdropCampaignChain(address _airdropCampaignAddress, Chains _airdropChain) external onlyAdmin {
+        if (_airdropChain == Chains.Zero) revert ChainDataNotSet();
+        if (airdrops[_airdropCampaignAddress].chain != Chains.Zero) revert ChainAlreadySet();
+        if (airdrops[_airdropCampaignAddress].airdropFinished == true) revert AirdropHasFinished();
```
no need for == true

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+        onlyAdmin
+    {
+        uint256 airdropChainArrayLength = _airdropChainArray.length;
+        if (airdropChainArrayLength > batchChainArrayLimit) revert ArrayTooLong();
```
next line covers this

nlukic97 replied:
Same a n previous example above, will discuss with teammates if we should remove the check for array length

nlukic97 replied:
Confirmed - removing the check.

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+    /// @notice Disabiling the ability to push hashes to the hashArray of a specific campaign
+    /// @param _airdropCampaignAddress - address of airdropCampaign contract
+    function finalizeAirdrop(address _airdropCampaignAddress) public onlyAdmin {
+        if (airdrops[_airdropCampaignAddress].airdropFinished == true) revert AlreadyFinalized();
```
remove == true

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+
+/// @title AirdropCampaignData - Data contract for storing of daily merkleRootHashes of airbro Campaigns
+contract AirdropCampaignData {
+    address public admin = 0xF4b5bFB92dD4E6D529476bCab28A65bb6B32EFb3;
```
constructor?

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+/// @title AirdropCampaignData - Data contract for storing of daily merkleRootHashes of airbro Campaigns
+contract AirdropCampaignData {
+    address public admin = 0xF4b5bFB92dD4E6D529476bCab28A65bb6B32EFb3;
+    uint16 public batchHashArrayLimit = 600;
```
remove this and next one, and remove checks, waste of gas.
Admin cal always simulate tx before sending it + metamask does that automatically

nlukic97 replied:
Yes. Now that I think about it, the backend can just check the length of the array before submitting, so I agree this check is unnecessary.

nlukic97 replied:
Checked with team - the variable will be removed.

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,130 @@
+        bool airdropFinished;
+    }
+
+    mapping(address => AirdropData) public airdrops;
```
Note: Smart contracts can have the same address on different chains. (CREATE2) With current implementation, only one chain could get the drop.

nlukic97 replied:
Not sure If I understand this

bxmmm1 replied:
> Not sure If I understand this.
> 
> EDIT: I see, so we can use the [CREATE2](https://github.com/thegostep/solidity-create2-deployer) to deploy on two chains with the same address. 

This means that other people can do the same thing as well. 
This means that you could reward same address only once (on one chain)
Maybe i misunderstood, but what is the point of having `enum Chains {
        Zero,
        Eth,
        Pols
    }`

nlukic97 replied:
This okay since we're using one chain per contract

nlukic97 commented:
Fixed many things, also removed some deprecated contracts. 
If necessary to access them again for any reason, this is the commit tree where they were last present:
https://github.com/MVPWorkshop/Airbro-contracts/tree/5806aa12c8caf1c600c533982341b52285fe948c

bxmmm1 commented on contracts/airdrops/Existing1155NftDrop.sol:
```diff
@@ -139,40 +101,33 @@ contract Existing1155NftDrop is AirdropInfo, AirdropMerkleProof, IERC1155Receive
+    /// @param tokenId the token id based on which the user wishes to claim the reward
+    /// @return bool if user is eligible for reward
+    function isEligibleForReward(uint256 tokenId) public view returns (bool) {
+        if ((block.timestamp > airdropFinishTime) || (hasClaimed[tokenId]) || (rewardedNft.ownerOf(tokenId) != msg.sender)) {
+            return false;
+        } else {
+            return true;
+        }
```
You dont need `else` block when you are returning inside of IF
```
if (something) {
    return false;
}
return true;
```

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -64,95 +48,76 @@ contract ExistingTokenDrop is AirdropInfo, AirdropMerkleProof, Ownable {
+    function isEligibleForReward(uint256 tokenId) public view returns (bool) {
+        if ((block.timestamp > airdropFinishTime) || (hasClaimed[tokenId]) || (rewardedNft.ownerOf(tokenId) != msg.sender)) {
+            return false;
+        } else {
```
no need for else

bxmmm1 commented on contracts/airdrops/ExistingTokenDrop.sol:
```diff
@@ -3,59 +3,43 @@ pragma solidity ^0.8.14;
 
 /// @title Airdrops existing ERC20 tokens for airdrop recipients
-contract ExistingTokenDrop is AirdropInfo, AirdropMerkleProof, Ownable {
+contract ExistingTokenDrop is AirdropMerkleProof {
```
Do you really need to extend `AirdropMerkleProof`? I don't see you using `checkProof` in this contract

nlukic97 replied:
You are right. I will remove them from all contracts that do not actually implement a MerkleProof check.

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,115 @@
+        merkleRoot = _merkleRoot;
+        airdropExpirationTimestamp = block.timestamp + 60 days;
+        numberOfClaimers = _numberOfClaimers;
+        tokensPerClaim = tokenSupply / numberOfClaimers;
```
use `_numberOfClaimers` instead of `numberOfClaimers` (cheaper)

nlukic97 replied:
For line 57 - good catch ! Updated

bxmmm1 commented on contracts/airdrops/shared/AirdropExistingToken.sol:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /// @notice Claim multiple ERC20 airdrops at once
+    /// @param tokenIds are the rewarded NFT collections token ID's
```
add @dev comment explaining that this function doesn't transfer token, and that it should be handled in child contract

bxmmm1 commented on contracts/airdrops/shared/AirdropExistingToken.sol:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /// @notice Allows the NFT holder to claim their ERC20 airdrop
+    /// @param tokenId is the rewarded NFT collections token ID
```
/// @dev explain that it doesn't contain logic for claiming (same for batchClaim)

bxmmm1 commented on contracts/campaignAirdrops/NewSB1155DropCampaign.sol:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    function burn() external {
+        if (balanceOf(msg.sender, _tokenId) != 1) revert NotTokenOwner();
```
_burn does this check internally

GotchaGuy replied:
Should I then leave it like this: 
```
function burn() external {
        _burn(msg.sender, _tokenId, _tokenAmount);
    }
```

or should I implement ERC1155Burnable @bxmmm1

GotchaGuy replied:
After solving this _burn comment is it okay to merge with main? @bxmmm1

bxmmm1 replied:
I did not read most recent .js files and tests, but solidity code looks almost ready. Minor changes are required to make it a bit better

bxmmm1 commented on contracts/campaignAirdrops/NewSB1155DropCampaign.sol:
```diff
@@ -0,0 +1,93 @@
+    /// @notice Checks if the user is eligible for this airdrop
+    /// @param _merkleProof is the merkleRoot proof that this user is eligible for claiming reward
+    /// @return true if user is eligibleto claim a reward
+    function isEligibleForReward(bytes32[] calldata _merkleProof) public view virtual override returns (bool) {
```
you aren't adding any extra logic in this override, you can remove it and it will be inherited from parent

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,100 @@
+        numberOfClaimers = _numberOfClaimers;
+        tokensPerClaim = tokenSupply / _numberOfClaimers;
+
+        emit MerkleRootSet(_merkleRoot, _numberOfClaimers);
```
This kinda belongs to parents `setMerkleRootHandler`. Here you are mixing apple and peaches.
Number of claimers has nothing to do with merkle root.
I'd emit `emit MerkleRootSet(_merkleRoot)` inside of `super.setMerkleRootHandler`, and remove it from child contracts.
For this particular use case maybe use different event? emit NumberOfClaimersSet or whatever

nlukic97 replied:
I thought about this, and I think you are right. Will update.

bxmmm1 commented on contracts/AirdropCampaignData.sol:
```diff
@@ -0,0 +1,127 @@
+
+    /// @notice Disabiling the ability to push hashes to the hashArray of a specific campaign
+    /// @param _airdropCampaignAddress - address of airdropCampaign contract
+    function finalizeAirdrop(address _airdropCampaignAddress) public onlyAdmin {
```
why public and not external?

bxmmm1 commented on contracts/campaignAirdrops/NewSB1155DropCampaign.sol:
```diff
@@ -0,0 +1,85 @@
+    function setMerkleRoot(bytes32 _merkleRoot) external onlyAdmin {
+        super.setMerkleRootHandler(_merkleRoot);
+
+        emit MerkleRootSet(_merkleRoot);
```
This event should be inside of paremt method, and then this method doesn't need to exist in a child contract

bxmmm1 commented on contracts/campaignAirdrops/NewERC1155DropCampaign.sol:
```diff
@@ -0,0 +1,46 @@
+    /// @param _merkleRoot - The root hash of the Merle Tree
+    function setMerkleRoot(bytes32 _merkleRoot) external onlyAdmin {
+        super.setMerkleRootHandler(_merkleRoot);
+        emit MerkleRootSet(_merkleRoot);
```
This event should be inside of paremt method, and then this method doesn't need to exist in a child contract

bxmmm1 commented on contracts/campaignAirdrops/ExistingERC20DropCampaign.sol:
```diff
@@ -0,0 +1,98 @@
+        numberOfClaimers = _numberOfClaimers;
+        tokensPerClaim = tokenSupply / _numberOfClaimers;
+
+        emit MerkleRootSet(_merkleRoot);
```
This event should be inside of paremt method
